### PR TITLE
Rely on ciso646 and __cplusplus macro when detecting cxx, add cxx11 flag

### PR DESCRIPTION
**Pull request by sovrasov**
It seems that `_LIBCPP_VERSION` is not defined in the `string` header at least for GCC 5.4 and above.
```
printf "#include <string>\nint main () {}" | g++ -E -x c++ -dM - | grep CPP
```

### CODE REVIEW

**Comment by sovrasov** [comment]
@jschueller please take a look. I need CXX11 macro in #194 
It doesn't affect current algorithms, but can be useful in future.

File: nlopt_config.h.in [hunk 1]
```diff
@@ -150,6 +150,9 @@
 /* Define if compiled including C++-based routines */
 #cmakedefine NLOPT_CXX
 
+/* Define if compiled including C++11-based routines */
+#cmakedefine NLOPT_CXX11
```
Comment (stevengj) [inline nlopt_config.h.in]: Is this actually used anywhere?  Why do we need `-std=c++11` at all?

Comment (stevengj) [reply nlopt_config.h.in]: The STOGO code is ancient — mostly written in 1998, and incorporated in NLopt in 2007.  How could it possibly depend heavily on C++11?

Comment (sovrasov) [reply nlopt_config.h.in]: I meant future code and the code from #194.

**Comment by sovrasov** [comment]
Moved to #194

**Pull request closed without merging**